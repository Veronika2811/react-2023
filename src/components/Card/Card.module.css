.card {
  position: relative;
  width: 250px;
  text-align: center;
  list-style: none;
  border: 1px solid #4d4b4b;
  border-radius: 10px;
  transition: border-color 0.3s;
  cursor: pointer;
}

.card:hover {
  border-color: #918e8e;
}

.card__image {
  width: 100%;
  height: 250px;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
}

.card__label {
  position: absolute;
  top: 10px;
  right: 10px;
  padding: 0px 10px;
  border-radius: 5px;
}

.card__label_red {
  background-color: #ec2d01;
}

.card__label_green {
  background-color: #76c076;
}

.card__label_grey {
  background-color: #808080;
}

.card__content {
  padding: 10px;
}
