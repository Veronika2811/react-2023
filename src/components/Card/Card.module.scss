.card {
  border: 1px solid #d8d8d8;
  border-radius: 10px;
  text-align: center;
  position: relative;
  width: 250px;

  &:hover {
    cursor: pointer;
  }

  &__item {
    font-size: 15px;
  }

  &__value {
    font-style: italic;
  }

  &__image {
    height: 250px;
    width: 100%;
    border-top-left-radius: 10px;
    border-top-right-radius: 10px;
  }

  &__label {
    position: absolute;
    right: 10px;
    top: 10px;
    color: white;
    font-weight: 700;
    padding: 5px;
    border-radius: 5px;
    font-size: 15px;
  }

  &__description {
    padding: 15px;
    display: flex;
    flex-direction: column;
    gap: 5px;
  }
}
