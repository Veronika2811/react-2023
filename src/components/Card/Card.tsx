import React from 'react';

import classes from './Card.module.scss';

interface ICardList {
  name: string;
  age: number;
  email: string;
  password: string;
  passwordConfirmation: string;
  gender: string;
  acceptCheckbox: boolean;
  image: string;
  country: string;
}

const Card = ({ card }: { card: ICardList }) => {
  return (
    <div className={classes.card}>
      <img className={classes.card__image} src={card.image} />
      <div className={classes.card__description}>
        <p className={classes.card_field}>
          Name: <span className={classes.card_value}>{card.name}</span>
        </p>
        <p className={classes.card_field}>
          Age: <span className={classes.card_value}>{card.age}</span>
        </p>
        <p className={classes.card_field}>
          Email: <span className={classes.card_value}>{card.email}</span>
        </p>
        <p className={classes.card_field}>
          Password:{' '}
          <span className={classes.card_value}>
            {card.password} / {card.passwordConfirmation}
          </span>
        </p>
        <p className={classes.card_field}>
          Gender: <span className={classes.card_value}>{card.gender}</span>
        </p>
        <p className={classes.card_field}>
          Country: <span className={classes.card_value}>{card.country}</span>
        </p>
      </div>
    </div>
  );
};

export default Card;
