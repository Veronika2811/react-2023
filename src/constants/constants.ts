export const BASE_URL = 'https://rickandmortyapi.com/api/character';

export const DEFAULT_VALUE_PER_PAGE = 20;
export const ADDITIONAL_VALUE_PER_PAGE = 10;

export const LOCAL_STORAGE_KEY = 'Veronika2811-react-2023__searchRequest';

export const DETAILS_URL_PARAMETER_KEY = 'details';
export const PAGE_URL_PARAMETER_KEY = 'page';

export const DEFAULT_PAGE = '1';
