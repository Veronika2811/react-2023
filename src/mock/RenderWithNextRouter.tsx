import React, { ReactNode } from 'react';
import { RouterContext } from 'next/dist/shared/lib/router-context.shared-runtime';

import { createMockRouter } from './createMockRouter';

interface IRenderWithNextRouter {
  queryParams?: { [key: string]: string };
  children: ReactNode;
}

const RenderWithNextRouter = ({
  queryParams = {},
  children,
}: IRenderWithNextRouter) => {
  const { Provider } = RouterContext;

  return (
    <Provider value={createMockRouter({ query: queryParams })}>
      {children}
    </Provider>
  );
};

export default RenderWithNextRouter;
