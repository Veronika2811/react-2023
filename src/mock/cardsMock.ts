export const MortyMock = {
  created: '2017-11-04T18:50:21.651Z',
  episode: ['https://rickandmortyapi.com/api/episode/1'],
  gender: 'Male',
  id: 2,
  image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
  location: {
    name: 'Citadel of Ricks',
    url: 'https://rickandmortyapi.com/api/location/3',
  },
  name: 'Morty Smith',
  origin: {
    name: 'unknown',
    url: '',
  },
  species: 'Human',
  status: 'Alive',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/2',
};

export const AlbertMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 11,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Albert Einstein',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};
