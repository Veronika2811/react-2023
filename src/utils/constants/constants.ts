export const BASE_URL = 'https://rickandmortyapi.com/api/character';

export const DEFAULT_PAGE = '1';

export const DEFAULT_VALUE_PER_PAGE = '20';
export const ADDITIONAL_VALUE_PER_PAGE = '10';

export const DEFAULT_QUERY_PARAMS = {
  currentPage: DEFAULT_PAGE,
  perPage: DEFAULT_VALUE_PER_PAGE,
};
