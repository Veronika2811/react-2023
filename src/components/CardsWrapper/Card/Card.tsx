import { useRouter } from 'next/router';
import Image from 'next/image';

import { IDataResult } from '@/types/types';
import getStatusCharacterColor from '@/utils/getStatusCharacterColor/getStatusCharacterColor';
import { DEFAULT_QUERY_PARAMS } from '@/utils/constants/constants';

import classes from './Card.module.css';

const Card = ({ card }: { card: IDataResult }) => {
  const router = useRouter();

  const { id, status, image, gender, species, name, location } = card;

  const openCardDetails = () => {
    const { name: query, page, perPage } = router.query;

    router.push({
      pathname: '/',
      query: {
        details: id,
        ...(query ? { name: query } : {}),
        page: page || DEFAULT_QUERY_PARAMS.currentPage,
        perPage: perPage || DEFAULT_QUERY_PARAMS.perPage,
      },
    });
  };

  return (
    <li className={classes.card} onClick={openCardDetails} data-testid="card">
      <p
        className={`${classes.card__label} ${
          classes[getStatusCharacterColor(status)]
        }`}
      >
        {status}
      </p>
      <Image src={image} alt={name} width={250} height={250} priority={true} />
      <div className={classes.card__content}>
        <h2>{name}</h2>
        <p>{`Gender: ${gender}`}</p>
        <p>{`Species: ${species}`}</p>
        <p>{`Location: ${location.name}`}</p>
      </div>
    </li>
  );
};

export default Card;
