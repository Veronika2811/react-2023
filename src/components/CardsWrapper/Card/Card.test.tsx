import { expect, it } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { RouterContext } from 'next/dist/shared/lib/router-context.shared-runtime';

import Card from './Card';
import { MortyMock } from '@/mock/cardsMock';
import RenderWithNextRouter from '@/mock/RenderWithNextRouter';
import { createMockRouter } from '@/mock/createMockRouter';
import { DEFAULT_QUERY_PARAMS } from '@/utils/constants/constants';

describe('Card component', () => {
  it('renders correctly Card component', () => {
    const container = render(
      <RenderWithNextRouter>
        <Card card={MortyMock} />
      </RenderWithNextRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it('should display relevant character data Card component', () => {
    render(
      <RenderWithNextRouter>
        <Card card={MortyMock} />
      </RenderWithNextRouter>
    );

    const { status, name, gender, species, location } = MortyMock;

    const STATUS = screen.getByText(status);
    expect(STATUS).toBeInTheDocument();

    const NAME = screen.getByText(name);
    expect(NAME).toBeInTheDocument();

    const GENDER = screen.getByText(`Gender: ${gender}`);
    expect(GENDER).toBeInTheDocument();

    const SPECIES = screen.getByText(`Species: ${species}`);
    expect(SPECIES).toBeInTheDocument();

    const LOCATION_NAME = screen.getByText(`Location: ${location.name}`);
    expect(LOCATION_NAME).toBeInTheDocument();
  });

  it('should preserve the existing name query params in the router and open the parts card when you click', () => {
    const router = createMockRouter({
      query: { name: MortyMock.name },
    });

    render(
      <RouterContext.Provider value={router}>
        <Card card={MortyMock} />
      </RouterContext.Provider>
    );

    const card = screen.getByTestId('card');
    fireEvent.click(card);

    expect(router.push).toBeCalledWith({
      pathname: '/',
      query: {
        details: MortyMock.id,
        name: MortyMock.name,
        page: DEFAULT_QUERY_PARAMS.currentPage,
        perPage: DEFAULT_QUERY_PARAMS.perPage,
      },
    });
  });

  it('should set the query parameters to default and open the parts card when you click', async () => {
    const router = createMockRouter({});

    render(
      <RouterContext.Provider value={router}>
        <Card card={MortyMock} />
      </RouterContext.Provider>
    );

    const card = screen.getByTestId('card');
    fireEvent.click(card);

    expect(router.push).toBeCalledWith({
      pathname: '/',
      query: {
        details: MortyMock.id,
        page: DEFAULT_QUERY_PARAMS.currentPage,
        perPage: DEFAULT_QUERY_PARAMS.perPage,
      },
    });
  });
});
