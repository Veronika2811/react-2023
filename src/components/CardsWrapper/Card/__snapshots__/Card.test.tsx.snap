// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > renders correctly Card component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="_card_f767b2"
        data-testid="card"
      >
        <p
          class="_card__label_f767b2 _card__label_green_f767b2"
        >
          Alive
        </p>
        <img
          alt="Morty Smith"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="250"
          src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75"
          srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75 2x"
          style="color: transparent;"
          width="250"
        />
        <div
          class="_card__content_f767b2"
        >
          <h2>
            Morty Smith
          </h2>
          <p>
            Gender: Male
          </p>
          <p>
            Species: Human
          </p>
          <p>
            Location: Citadel of Ricks
          </p>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="_card_f767b2"
      data-testid="card"
    >
      <p
        class="_card__label_f767b2 _card__label_green_f767b2"
      >
        Alive
      </p>
      <img
        alt="Morty Smith"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="250"
        src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75"
        srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75 2x"
        style="color: transparent;"
        width="250"
      />
      <div
        class="_card__content_f767b2"
      >
        <h2>
          Morty Smith
        </h2>
        <p>
          Gender: Male
        </p>
        <p>
          Species: Human
        </p>
        <p>
          Location: Citadel of Ricks
        </p>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
